src/Data/ConcurrentResourceMap.hs

```diff
@@ -162,6 +164,17 @@ withInitialisedResource vm k destroyResource act = bracket
         CountedResource { resource = Initialised r } -> act (Just r)
         _ -> act Nothing
 
+-- | Perform an action for each resource. Useful for destroying all the resources
```

Comment:
Sadly I think this is completely the opposite of what this library tries to provide. `withSharedResource` is basically the only entry point and what it ensures in a concurrent setting is:

* initialiser runs only if resource is not initialised already and there's no race
* there can be many different resources in single map and there's no real contention for separate resources
* deinitialisation runs when last user exits and there's no race

(small caveat: we don't protect against leaking resources via IORef etc so I'm going to be ignoring that)

This is done by some counting of users.

What this added function does is:

* it locks the whole map for duration of its run (no new users are allowed to come)
* it accesses resource without registering itself as user so another thread can initialise/deinitialise the resource while it's "in use": note that locking the whole map at start doesn't do anything positive: all other users have references to the MVars at the values of the map and are modifying those directly, only going back to the original map once they are done to clean it up. Taking top mvar blocks any users that are first or last from making progress (but doesn't block them from init/deinit so this is not useful at all, just harmful)
* the docs hint at destroying resources which really talks about modifying underlying resource (`r`) while so far in `withSharedResource` we've been using this language to talk about deinitialising resources. So it's probably a bit confusing. Indeed, we're not deinitialising anything at least as the library is concerned.

The question is what to do. I think the _closest_ thing that gets what you want is:

```haskell
keys <- readMVar vm
forM_ keys $ \key -> withInitialisedResource key <deinitialiser> act
```

It ends up with similar result to what you we already had here but one that probably doesn't achieve what we wanted which is to perform something on _all_ the resources. If we want to block new/old users from interactive with resources, we'd have to change the API of `withSharedResource` to express failure or throw new exceptions or something.

I think what I would suggest and the actual intended way to use the lib is to get your users to terminate and exit properly rather than having a backdoor that terminates the resource and presumably makes the users die that way. It's just the wrong way around from what was intended here.

Let me know if I'm mistaken somehow or the lib needs to be redesigned but I think the proposed function is probably just wrong and I don't know how to make it right in such a way that it's still useful for doing something on all resources, at least without thinking about it a bit.

Reply:
> the closest thing that gets what you want is:

No, I also want to prevent new resources from being created. I think I now understand the problem you pointed out; I have to think a bit more.

If we were to make `closeResourceMap`, I guess throwing an exception from `withSharedResource` makes more sense when the resource map is already closed

Reply:
I guess we can do something in userland like this:

* make a new ref `TVar (Maybe Int)` that starts as `Just 0`
* in initialiser, before creating new resource check that it's `Just` and increase by 1. Create resource. In case it's Nothing, throw an exception to jump out or just yield the resource in some closed state and do nothing with it in the callback.
* when we want to "close" the map, we wait for TVar to become Just 0 and change it to Nothing atomically. 

This is a way to ensure that we either never create a new value in map (exception way) or we create one but the value is not useful (some closed value, say Nothing).

Once we've done this, we can get the list of all existing keys from the map and we have guarantee that they are all the interesting keys (they may not be all keys necessarily if we're going with "closed" resource way which is fine). Once we have all interesting keys, we map each one with withInitialisedResouce and modify it as we please, presumably signaling that it should stop being used. All the users then die/disconnect/whatever and last one deinitialised the resource "proper" which may be a noop on the resource at this point and just results in removal from the map.

All this is in userland. Sounds a bit complicated when typed out but it's simple in practice I think, at least the non-exception way. It also doesn't lock up the whole map which is good.

Reply:
I guess to facilitate this we just need to add a function to get all the current keys from the map, with documentation about how it's probably only useful with some external synchronisation happening.

Reply:
OK, I see. I changed the PR to add `keysResourceMap :: ResourceMap m => ConcurrentResourceMap m v -> IO [Key m]`, along with an addition to the `ResourceMap` class (a breaking change).